src/hotspot/share/gc/shenandoah/heuristics/shenandoahHeuristics.hpp

```diff
@@ -58,6 +58,26 @@ class ShenandoahCollectionSet;
 class ShenandoahHeapRegion;
 class ShenandoahGeneration;
 
+// TODO: Probably want to subclass ShenandoahHeuristics.
+//
+// As currently implemented, we have three instances of
+// ShenandoahHeuristics, for GLOBAL, OLD, and YOUNG.  Each instance
+// needs different information.
+//
+// In particular:
+//  YOUNG needs a pointer to OLD so it can ask if there are OLD regions
+//   wanting to be evacuated.
+//  OLD needs to keep track of which OLD regions need to be evacuated and
+//   which OLD regions need their garbage to be coalesced and filled.
+//   OLD also needs to know when all of its candidate evacuation regions
+//   have been evacuated because it cannot trigger a new OLD
+//   concurrent mark until that has happened.
+//  GLOBAL: not sure what this needs, or how it interacts with OLD, YOUNG
+//   heuristics.
+//
+// For expediency, we'll have a single class that is a "union" of
+// necessary functionality.
+
```

Comment:
Hmmhmm, this seems a suboptimal design. My intuition would have been to not have each heuristic need to know about any old heuristic, but keep them separate, and compose functionality elsewhere, but I leave that to you. The way it looks now it may end up recursive (young-heuristics->old-heuristics->old-heuristics->...-> etc)

Reply:
This has been addressed with recently committed refactoring of ShenandoahOldHeuristics and subclasses.

Reply:
This has been addressed with recently pushed refactoring of ShenandoahOldHeuristics and subclasses.